src/builder.rs

```diff
@@ -780,45 +969,78 @@ fn build_bundle<B, R: RngCore>(
-            .collect::<Vec<_>>();
+        // Use Vec::with_capacity().extend(...) instead of .collect() to avoid reallocations,
+        // as we can estimate the vector size beforehand.
+        let mut indexed_spends_outputs =
+            Vec::with_capacity(spends.len().max(outputs.len()).max(MIN_ACTIONS));
 
-        let mut indexed_outputs = outputs
-            .into_iter()
-            .chain(iter::repeat_with(|| OutputInfo::dummy(&mut rng)))
-            .enumerate()
-            .take(num_actions)
-            .collect::<Vec<_>>();
+        let spends_outputs_by_asset = partition_by_asset(&spends, &outputs);
+
+        indexed_spends_outputs.extend(spends_outputs_by_asset.into_iter().flat_map(
+            |(asset, (spends, outputs))| {
+                let num_asset_pre_actions = spends.len().max(outputs.len());
+
+                let first_spend = spends.first().map(|(s, _)| s.clone());
+
+                let mut indexed_spends = spends
+                    .into_iter()
+                    .chain(iter::repeat_with(|| {
+                        (
+                            pad_spend(first_spend.as_ref(), asset, &mut rng)
+                                .unwrap_or_else(|err| panic!("{:?}", err)),
+                            None,
+                        )
+                    }))
+                    .take(num_asset_pre_actions)
+                    .collect::<Vec<_>>();
+
+                let mut indexed_outputs = outputs
+                    .into_iter()
+                    .chain(iter::repeat_with(|| {
+                        (OutputInfo::dummy(&mut rng, asset), None)
+                    }))
+                    .take(num_asset_pre_actions)
+                    .collect::<Vec<_>>();
 
-        // Shuffle the spends and outputs, so that learning the position of a
-        // specific spent note or output note doesn't reveal anything on its own about
-        // the meaning of that note in the transaction context.
-        indexed_spends.shuffle(&mut rng);
-        indexed_outputs.shuffle(&mut rng);
+                // Shuffle the spends and outputs, so that the position does not reveal any
+                // information about the content.
+                indexed_spends.shuffle(&mut rng);
+                indexed_outputs.shuffle(&mut rng);
+
+                assert_eq!(indexed_spends.len(), indexed_outputs.len());
+
+                indexed_spends.into_iter().zip(indexed_outputs)
+            },
+        ));
+
+        indexed_spends_outputs.extend(
+            iter::repeat_with(|| {
+                (
+                    (SpendInfo::dummy(AssetBase::native(), &mut rng), None),
+                    (OutputInfo::dummy(&mut rng, AssetBase::native()), None),
+                )
+            })
+            .take(MIN_ACTIONS.saturating_sub(indexed_spends_outputs.len())),
+        );
+
+        // Shuffle the spends and outputs, so that the position does not reveal any information
+        // about the content.
```

Comment:
_From str4d ([r2602541601](https://github.com/zcash/orchard/pull/471#discussion_r2602541601))_

```suggestion
        // We shuffled the spends and outputs within each `AssetBase` above; now we
        // shuffle the actions to achieve a similar property across `AssetBase`s.
```

Reply:
FIXED